---
title: "R Assignment"
author: "Dylan Harris"
date: "2023-03-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Data Inspection
```{r}
library(tidyverse)
genotypes <- read_tsv("https://github.com/dah96/bcb546/raw/master/R_Assignment/fang_et_al_genotypes.txt")
snp_positions <- read_tsv("https://github.com/dah96/bcb546/raw/master/R_Assignment/snp_position.txt")
```
Loaded the tidyverse package and loaded the two data files as dataframes in R.
```{r}
nrow(genotypes)
ncol(genotypes)
object.size(genotypes)
```
fang_et_al_genotypes.txt has 2782 rows, 986 columns, and is 23.1Mb in size.
```{r}
nrow(snp_positions)
ncol(snp_positions)
object.size(snp_positions)
```
snp_position.txt is 983 rows, 15 columns, and is 359kb is size.

# Data Processing
```{r}
maize_genotypes <- filter(genotypes, Group %in% c("ZMMIL","ZMMLR","ZMMMR"))
teosinte_genotypes <- filter(genotypes, Group %in% c("ZMPBA","ZMPJA","ZMPIL"))
```
Filtered genotypes into maize and teosinte genotypes respectively.
```{r}
t_m_genotypes <- t(maize_genotypes) %>%
  data.frame()
colnames(t_m_genotypes) <- t_m_genotypes[c(1),]
t_m_genotypes <- t_m_genotypes[-c(1,2,3),]
t_t_genotypes <-  t(teosinte_genotypes) %>%
  data.frame()
colnames(t_t_genotypes) <- t_t_genotypes[c(1),]
t_t_genotypes <- t_t_genotypes[-c(1,2,3),]
```
Transposes the maize genotypes, but doing so converts the data frame into a matrix, which we turn back into a data frame. The column names also need to be adjusted to the first row of the transposed data frame, then the superfluous rows are removed. Same process is repeated for teosinte genotypes.
```{r}
snp_cut <- select(snp_positions, SNP_ID, Chromosome, Position)
maize_joined <- bind_cols(snp_cut, t_m_genotypes)
teosinte_joined <-  bind_cols(snp_cut, t_t_genotypes)
```
snp_positions is used to create a new dataframe only containing SNP_ID, Chromosome, and Positions columns which is then joined with the transposed maize and teosinte genotypes respectively.
```{r}
#Almost working for loop, needs to be run as maize_chr[i] first then maize_chr[[i]],
#makes a list of lists when done
#maize_chr <- vector(length = 10)
#for (i in 1:10) {
#  maize_chr[[i]] <- filter(maize_joined, Chromosome == i) %>%
#    arrange(as.numeric(Position))
#}
#maize_chr[[1]] <- filter(maize_joined, Chromosome == 1) %>%
#  arrange(as.numeric(Position))
maize_chr1 <- filter(maize_joined, Chromosome == 1) %>%
  arrange(as.numeric(Position))
maize_chr2 <- filter(maize_joined, Chromosome == 2) %>%
  arrange(as.numeric(Position))
maize_chr3 <- filter(maize_joined, Chromosome == 3) %>%
  arrange(as.numeric(Position))
maize_chr4 <- filter(maize_joined, Chromosome == 4) %>%
  arrange(as.numeric(Position))
maize_chr5 <- filter(maize_joined, Chromosome == 5) %>%
  arrange(as.numeric(Position))
maize_chr6 <- filter(maize_joined, Chromosome == 6) %>%
  arrange(as.numeric(Position))
maize_chr7 <- filter(maize_joined, Chromosome == 7) %>%
  arrange(as.numeric(Position))
maize_chr8 <- filter(maize_joined, Chromosome == 8) %>%
  arrange(as.numeric(Position))
maize_chr9 <- filter(maize_joined, Chromosome == 9) %>%
  arrange(as.numeric(Position))
maize_chr10 <- filter(maize_joined, Chromosome == 10) %>%
  arrange(as.numeric(Position))
```
Maize chromosomes pulled from joined data frame via filter and arranged via numerical SNP  position.
```{r}
maize_alt_joined <- data.frame(lapply(maize_joined, function(x) {
  gsub("[?]", "-", x)
}))
maize_alt_chr1 <- filter(maize_alt_joined, Chromosome == 1) %>%
  arrange(desc(as.numeric(Position)))
maize_alt_chr2 <- filter(maize_alt_joined, Chromosome == 2) %>%
  arrange(desc(as.numeric(Position)))
maize_alt_chr3 <- filter(maize_alt_joined, Chromosome == 3) %>%
  arrange(desc(as.numeric(Position)))
maize_alt_chr4 <- filter(maize_alt_joined, Chromosome == 4) %>%
  arrange(desc(as.numeric(Position)))
maize_alt_chr5 <- filter(maize_alt_joined, Chromosome == 5) %>%
  arrange(desc(as.numeric(Position)))
maize_alt_chr6 <- filter(maize_alt_joined, Chromosome == 6) %>%
  arrange(desc(as.numeric(Position)))
maize_alt_chr7 <- filter(maize_alt_joined, Chromosome == 7) %>%
  arrange(desc(as.numeric(Position)))
maize_alt_chr8 <- filter(maize_alt_joined, Chromosome == 8) %>%
  arrange(desc(as.numeric(Position)))
maize_alt_chr9 <- filter(maize_alt_joined, Chromosome == 9) %>%
  arrange(desc(as.numeric(Position)))
maize_alt_chr10 <- filter(maize_alt_joined, Chromosome == 10) %>%
  arrange(desc(as.numeric(Position)))
```

Joined maize genotypes file had all "?" replaced with "-". Maize chromosomes are then pulled from the new alternative data frame and arranged in descending order of numerical SNP position.
```{r}
teosinte_chr1 <- filter(teosinte_joined, Chromosome == 1) %>%
  arrange(as.numeric(Position))
teosinte_chr2 <- filter(teosinte_joined, Chromosome == 2) %>%
  arrange(as.numeric(Position))
teosinte_chr3 <- filter(teosinte_joined, Chromosome == 3) %>%
  arrange(as.numeric(Position))
teosinte_chr4 <- filter(teosinte_joined, Chromosome == 4) %>%
  arrange(as.numeric(Position))
teosinte_chr5 <- filter(teosinte_joined, Chromosome == 5) %>%
  arrange(as.numeric(Position))
teosinte_chr6 <- filter(teosinte_joined, Chromosome == 6) %>%
  arrange(as.numeric(Position))
teosinte_chr7 <- filter(teosinte_joined, Chromosome == 7) %>%
  arrange(as.numeric(Position))
teosinte_chr8 <- filter(teosinte_joined, Chromosome == 8) %>%
  arrange(as.numeric(Position))
teosinte_chr9 <- filter(teosinte_joined, Chromosome == 9) %>%
  arrange(as.numeric(Position))
teosinte_chr10 <- filter(teosinte_joined, Chromosome == 10) %>%
  arrange(as.numeric(Position))
```
Teosinte chromosomes pulled from joined data frame via filter and arranged via numerical position.
```{r}
teosinte_alt_joined <- data.frame(lapply(teosinte_joined, function(x) {
  gsub("[?]", "-", x)
}))
teosinte_alt_chr1 <- filter(teosinte_alt_joined, Chromosome == 1) %>%
  arrange(as.numeric(Position))
teosinte_alt_chr2 <- filter(teosinte_alt_joined, Chromosome == 2) %>%
  arrange(as.numeric(Position))
teosinte_alt_chr3 <- filter(teosinte_alt_joined, Chromosome == 3) %>%
  arrange(as.numeric(Position))
teosinte_alt_chr4 <- filter(teosinte_alt_joined, Chromosome == 4) %>%
  arrange(as.numeric(Position))
teosinte_alt_chr5 <- filter(teosinte_alt_joined, Chromosome == 5) %>%
  arrange(as.numeric(Position))
teosinte_alt_chr6 <- filter(teosinte_alt_joined, Chromosome == 6) %>%
  arrange(as.numeric(Position))
teosinte_alt_chr7 <- filter(teosinte_alt_joined, Chromosome == 7) %>%
  arrange(as.numeric(Position))
teosinte_alt_chr8 <- filter(teosinte_alt_joined, Chromosome == 8) %>%
  arrange(as.numeric(Position))
teosinte_alt_chr9 <- filter(teosinte_alt_joined, Chromosome == 9) %>%
  arrange(as.numeric(Position))
teosinte_alt_chr10 <- filter(teosinte_alt_joined, Chromosome == 10) %>%
  arrange(as.numeric(Position))

```
Joined teosinte genotypes file had all "?" replaced with "-". Teosinte chromosomes are then pulled from the new alternative data frame and arranged in descending order of numerical SNP position.

# Data Visualization

##SNPs per Chromosome

```{r}
#Editing data for downstream processing
t_teo_mai_genotypes <- filter(genotypes, Group %in% c("ZMMIL","ZMMLR","ZMMMR","ZMPBA","ZMPJA","ZMPIL")) %>% 
  t() %>% 
  data.frame()
colnames(t_teo_mai_genotypes) <- t_teo_mai_genotypes[c(1),]
t_teo_mai_genotypes <- t_teo_mai_genotypes[-c(1,2,3),]
complete_join <- bind_cols(snp_cut, t_teo_mai_genotypes)
complete_join <- data.frame(lapply(complete_join, function(x) {
  gsub("[?]", "-", x)
}))
complete_pivot <- pivot_longer(complete_join, cols = !1:3, names_to = "Sample_ID")
names(complete_pivot)[names(complete_pivot) == "value"] <- "Genotype"
complete_pivot$heterozygosity <- ifelse(complete_pivot$Genotype %in% c("A/A", "T/T", "G/G", "C/C"), "Homo",
                                        ifelse(complete_pivot$Genotype == c("-/-"), "Missing", "Hetero"))
complete_pivot$Position <- as.numeric(complete_pivot$Position)
maize_pivot <- pivot_longer(maize_joined, cols = !1:3, names_to = "Sample_ID")
names(maize_pivot)[names(maize_pivot) == "value"] <- "Genotype"
maize_pivot$heterozygosity <- ifelse(maize_pivot$Genotype %in% c("A/A", "T/T", "G/G", "C/C"), "Homo",
                                        ifelse(maize_pivot$Genotype == c("-/-"), "Missing", "Hetero"))
maize_pivot$Position <- as.numeric(maize_pivot$Position)
teosinte_pivot <- pivot_longer(teosinte_joined, cols = !1:3, names_to = "Sample_ID")
names(teosinte_pivot)[names(teosinte_pivot) == "value"] <- "Genotype"
teosinte_pivot$heterozygosity <- ifelse(teosinte_pivot$Genotype %in% c("A/A", "T/T", "G/G", "C/C"), "Homo",
                                        ifelse(teosinte_pivot$Genotype == c("-/-"), "Missing", "Hetero"))
teosinte_pivot$Position <- as.numeric(teosinte_pivot$Position)
```
Created a teosinte + maize genotypes file to examine SNP position based on chromosome across both maize and teosinte. Pivoted complete, maize, and teosinte tables to put all genotype data in one column for easier analysis downstream and accurate number of total SNPs.
```{r}
ggplot(data = complete_pivot) +
  geom_bar(mapping = aes(x=(Chromosome), fill = Chromosome), show.legend = FALSE)
```

Bar Graph of number of SNPs per chromosome across both maize and teosinte genotypes. Amount of SNPs per chromosome appears to decrease for the most apart as the chromosome number increases, exceptions being chromosome 5 and 7. This is likely related to the size of the chromosomes decreasing from 1 to 10.

```{r}
ggplot(data = maize_pivot) +
  geom_point(mapping = aes(x=Chromosome, y=Position), color = "purple") +
  ggtitle("Maize SNPs by Chromosome")

ggplot(data = teosinte_pivot) +
  geom_point(mapping = aes(x=Chromosome, y=Position), color = "green") +
  ggtitle("Teosinte SNPs by Chromosome")
```

Dot plots comparing SNPs distribution on chromosomes between maize and teosinte. SNPs appear to be less common within the center of each chromosome.


##Missing data and amount of heterozygosity

```{r}
ggplot(data = complete_pivot) +
  geom_bar(mapping = aes(x = Chromosome, fill = heterozygosity), position = "fill")
```

Bar graph comparing amount of heterozygous, homozygous, and missing sites across chromosomes of both teosinte and maize. Most sites are homozygous with about 12.5% being heterozygous and 5% being missing data across all chromosomes.

##My own visualization

```{r}
ggplot(data = maize_pivot) +
  geom_bar(mapping = aes(x=Chromosome, fill=Genotype), position="fill") +
  ggtitle("Maize Genotypes")
ggplot(data = teosinte_pivot) +
  geom_bar(mapping = aes(x=Chromosome, fill=Genotype), position="fill") +
  ggtitle("Teosonite Genotypes")
```

Bar graphs comparing proportions of each genotype with missing data across both maize and teosinte samples. Each chromosome appears to have a differing amount of GC content from each other across each chromosome. Maize individuals also seem to have more instances of heterozygosity in there chromosomes.